### PR TITLE
home: Prompt for auxiliary groups in homectl firstboot

### PR DESCRIPTION
**Pull request by daandemeyer**

### CODE REVIEW

File: src/home/homectl.c [hunk 1]
```diff
@@ -2474,6 +2475,42 @@ static int create_interactively(void) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set userName field: %m");
 
+        for (;;) {
```
Comment (poettering) [inline src/home/homectl.c:2526]: hmm, I think we really should show a list of available groups first. i.e. enumerate all groups with groupdb_all(), suppress entries that do not have a disposition of either USER_SYSTEM or USER_REGULAR.

And then maybe finally suppress all groups for which identically named users exist and where both dispositions are USER_REGULAR.

Those restrictions should just be made when listing suggested groups. If user types in anything not listed we should just accept that.

Comment (poettering) [reply src/home/homectl.c:2526]: maybe generalize the show_menu() code from firstboot.c

**Comment by yuwata** [comment]
Please rebase.

File: src/home/homectl.c [hunk 2]
```diff
@@ -2476,6 +2476,109 @@ static int create_interactively(void) {
         if (r < 0)
                 return log_error_errno(r, "Failed to set userName field: %m");
 
+        _cleanup_(userdb_iterator_freep) UserDBIterator *iterator = NULL;
+        _cleanup_strv_free_ char **available = NULL, **groups = NULL;
+
+        r = groupdb_all(USERDB_SUPPRESS_SHADOW|USERDB_EXCLUDE_DYNAMIC_USER, &iterator);
+        if (r == -ENOLINK)
+                log_debug_errno(r, "No entries found. (Didn't check via Varlink.)");
```
Comment (poettering) [inline src/home/homectl.c]: maybe "groups" instead of "entries" here

File: src/home/homectl.c [hunk 3]
```diff
@@ -2476,6 +2476,109 @@ static int create_interactively(void) {
...
+        else if (r == -ESRCH)
+                log_debug_errno(r, "No entries found.");
```
Comment (poettering) [inline src/home/homectl.c]: similar

File: src/home/homectl.c [hunk 4]
```diff
@@ -2476,6 +2476,109 @@ static int create_interactively(void) {
...
+        else if (r < 0)
+                return log_error_errno(r, "Failed to enumerate groups: %m");
```
Comment (poettering) [inline src/home/homectl.c]: given this is not a key feature I'd just do log_warning_errno() and continue

File: src/home/homectl.c [hunk 5]
```diff
@@ -2463,7 +2463,7 @@ static int create_interactively(void) {
                         continue;
                 }
 
-                r = userdb_by_name(username, USERDB_SUPPRESS_SHADOW, /* ret= */ NULL);
+                r = userdb_by_name(username, USERDB_SUPPRESS_SHADOW|USERDB_EXCLUDE_DYNAMIC_USER, /* ret= */ NULL);
```
Comment (poettering) [inline src/home/homectl.c]: why add this flag? i think we should check for conflicts against dynamic users too

File: src/home/homectl.c [hunk 6]
```diff
@@ -2476,6 +2476,109 @@ static int create_interactively(void) {
...
+                                if (r >= 0 && user_record_disposition(ur) == USER_REGULAR)
+                                        continue;
+
+                        }
+
+                        r = strv_extend(&available, gr->group_name);
+                        if (r < 0)
+                                return log_oom();
+                }
+        }
```
Comment (poettering) [inline src/home/homectl.c]: would prefer if this piece would become its own function, i.e. acquire_group_list() or so

File: src/home/homectl.c [hunk 7]
```diff
@@ -2476,6 +2476,109 @@ static int create_interactively(void) {
...
+                               "%s Please enter an auxiliary group for user %s (empty to continue, \"list\" to list available groups): ",
+                               special_glyph(SPECIAL_GLYPH_TRIANGULAR_BULLET), username);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to query user for auxiliary group: %m");
+
+                if (isempty(s))
+                        break;
+
+                if (streq(s, "list")) {
+                        r = show_menu(available, /*n_columns=*/ 3, /*width=*/ 20, /*percentage=*/ 60);
```
Comment (poettering) [inline src/home/homectl.c:2550]: given that acquiring the group list is potentially kinda slow, maybe only acquire it here if not acquired yet.

File: src/home/homectl.c [hunk 8]
```diff
@@ -2424,6 +2424,55 @@ static int has_regular_user(void) {
...
+                        if (group_record_disposition(gr) == USER_REGULAR) {
+                                _cleanup_(user_record_unrefp) UserRecord *ur = NULL;
+
+                                r = userdb_by_name(gr->group_name, USERDB_SUPPRESS_SHADOW|USERDB_EXCLUDE_DYNAMIC_USER, &ur);
+                                if (r < 0 && r != -ESRCH)
+                                        return log_debug_errno(r, "Failed to check if matching user exists for group '%s': %m", gr->group_name);
+
+                                if (r >= 0 && user_record_disposition(ur) == USER_REGULAR)
+                                        continue;
+
```
Comment (poettering) [inline src/home/homectl.c]: spurius empty line

File: src/home/homectl.c [hunk 9]
```diff
@@ -2424,6 +2424,55 @@ static int has_regular_user(void) {
...
+                                break;
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed acquire next group: %m");
+
+                        if (!IN_SET(group_record_disposition(gr), USER_REGULAR, USER_SYSTEM))
+                                continue;
+
+                        if (group_record_disposition(gr) == USER_REGULAR) {
+                                _cleanup_(user_record_unrefp) UserRecord *ur = NULL;
+
```
Comment (poettering) [inline src/home/homectl.c:2456]: maybe a comment here: `/* Filter groups here that belong to a specific user, and are named like them */`

File: src/home/homectl.c [hunk 10]
```diff
@@ -2424,6 +2424,55 @@ static int has_regular_user(void) {
...
+                                r = userdb_by_name(gr->group_name, USERDB_SUPPRESS_SHADOW|USERDB_EXCLUDE_DYNAMIC_USER, &ur);
+                                if (r < 0 && r != -ESRCH)
+                                        return log_debug_errno(r, "Failed to check if matching user exists for group '%s': %m", gr->group_name);
+
+                                if (r >= 0 && user_record_disposition(ur) == USER_REGULAR)
```
Comment (poettering) [inline src/home/homectl.c]: maybe one more condition check:

`ur->gid == gr->gid`

i.e. only if the primary group of the user is the group we are looking at we skip the group.

File: src/home/homectl.c [hunk 11]
```diff
@@ -2424,6 +2424,55 @@ static int has_regular_user(void) {
...
+
+                        }
+
+                        r = strv_extend(&groups, gr->group_name);
+                        if (r < 0)
+                                return log_oom();
+                }
+        }
+
+        *ret = TAKE_PTR(groups);
```
Comment (poettering) [inline src/home/homectl.c:2475]: no strv_sort() here?

File: src/home/homectl.c [hunk 12]
```diff
@@ -2476,6 +2525,84 @@ static int create_interactively(void) {
...
+                if (r < 0)
+                        return log_error_errno(r, "Failed to check if specified group '%s' already exists: %m", s);
+
+                log_info("Selected '%s'.", s);
+
+                r = strv_extend(&groups, s);
+                if (r < 0)
+                        return log_oom();
+        }
+
```
Comment (poettering) [inline src/home/homectl.c:2595]: strv_sort_uniq() on the resulting list? after all, people might enter the same group three times, and we should clean that up.

File: src/home/homectl.c [hunk 13]
```diff
@@ -2603,6 +2603,45 @@ static int create_interactively(void) {
...
+                        log_info("No data entered, skipping.");
+                        break;
+                }
+
+                if (!valid_shell(shell)) {
+                        log_notice("Specified shell is not a valid UNIX shell path, try again: %s", shell);
+                        continue;
+                }
+
+                r = RET_NERRNO(access(shell, F_OK));
```
Comment (poettering) [inline src/home/homectl.c]: X_OK?

File: src/home/homectl.c [hunk 14]
```diff
@@ -2485,12 +2485,6 @@ static int create_interactively(void) {
 
         any_key_to_proceed();
 
-        r = acquire_bus(&bus);
```
Comment (poettering) [inline src/home/homectl.c:2439]: hmm, but this means the `_cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;` line above can be dropped too, no?

**Pull request merged**